WordDistanceInSpark.scala

```diff
@@ -1 +1,241 @@
+  at org.apache.spark.ContextCleaner$$anonfun$org$apache$spark$ContextCleaner$$keepCleaning$1.apply(ContextCleaner.scala:144)
+  at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1618)
+  at org.apache.spark.ContextCleaner.org$apache$spark$ContextCleaner$$keepCleaning(ContextCleaner.scala:143)
+  at org.apache.spark.ContextCleaner$$anon$3.run(ContextCleaner.scala:65)
+ERROR Utils: Uncaught exception in thread SparkListenerBus
+java.lang.InterruptedException
+  at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:998)
+  at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
+  at java.util.concurrent.Semaphore.acquire(Semaphore.java:312)
+  at org.apache.spark.util.AsynchronousListenerBus$$anon$1$$anonfun$run$1.apply$mcV$sp(AsynchronousListenerBus.scala:62)
+  at org.apache.spark.util.AsynchronousListenerBus$$anon$1$$anonfun$run$1.apply(AsynchronousListenerBus.scala:61)
+  at org.apache.spark.util.AsynchronousListenerBus$$anon$1$$anonfun$run$1.apply(AsynchronousListenerBus.scala:61)
+  at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1618)
+  at org.apache.spark.util.AsynchronousListenerBus$$anon$1.run(AsynchronousListenerBus.scala:60)*/
+
+
+  
+  
+  
+  type Hit = (Long, String) // (index, the significant word found at index)
+  type SameWordInterval = ((Long,Long),String) // ( (startIdx, endIdx), significant word)
+  
+  /**
+   * Finds the distance between given words in the text, defined as 
+   * minimum amount of other words in between occurrences of the given words.
+   * 
+   * Distance between a word and itself is 0.
+   * If one of the words does not occur in the text, distance is None
+   *  
+   * @param textFilePath
+   * @param word1
+   * @param word2
+   * @param sc sparkContext
+   * @return
+   */
+  def wordDistance(textFilePath:String, word1:String, word2:String)(implicit sc:SparkContext):Option[Long] = {
+    if (word1 == word2) return Some(0)
+    if (word1==null || word2==null) return None
+    
+    // need access to a cluster to tweak this experimentally, so this is just a guess. 
+    // I don't know whether even to fiddle with this at all or just take the defaultParallelism.
+    implicit val nbPartitions = sc.defaultParallelism * 2
+    
+    // would this normally have to be broadcasted to every node then?
+    val lines:RDD[String] = sc.textFile(textFilePath, nbPartitions)
+    
+    // look at text file as sequence of words
+    val words = asFlatWordList(lines)
+    
+    // we filter out all words that are not word1 or word2
+    val orderedHits:RDD[Hit] = findHitsInOrder(words, word1, word2)
+    
+    val smallestDistance:Option[Long] = findSmallestDistanceBetweenDifferentWordHits(orderedHits)
+    smallestDistance
+  }
+  
+  def asFlatWordList(lines:RDD[String]):RDD[String] = lines.flatMap { l => l.split(" ") }
```

Comment:
The subjective value added here that I intended was to make sure that in the top-level method wordDistance(), I don't get a mix of direct code and calls to other methods. Of course, my head now being a bit fresher again compared to Wednesday night or very early Thursday, it is clear that this would have been much nicer if I had actually followed it through. The reading of the textfile in line 59, should be folded into this method as well. Then there would be consistency, and not quite a one-liner.